### PR TITLE
[Feature] Add MonoFlex data augmentation label generation function

### PR DESCRIPTION
**Pull request by ZCMax**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Add monoflex pipeline 

## Modification
Add monoflex pipeline

## BC-breaking (Optional)
no

## Use cases (Optional)
n/a

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/datasets/pipelines/loading.py [hunk 1]
```diff
@@ -683,3 +684,343 @@ def __repr__(self):
...
+        pad_mode (str): the pad mode of image.
+            Default: default
+    """
+
+    def __init__(self, pad_mode='default'):
+        self.pad_mode = pad_mode
+
+    def __call__(self, results):
+        """Call function to generate edge indices for image.
+        Args:
```
Comment (wHao-Wu) [inline mmdet3d/datasets/pipelines/loading.py:703]: Is it necessary to insert a blank line between the summary and the args?

File: mmdet3d/datasets/pipelines/loading.py [hunk 2]
```diff
@@ -683,3 +684,343 @@ def __repr__(self):
...
+        use_local_coords (list[float]): Whether to use local coordinate of
+            keypooints. Default: True.
+    """
+
+    def __init__(self, use_local_coords=True):
+        self.use_local_coords = use_local_coords
+
+    def __call__(self, results):
+        """Call function to generate keypoints and its visible mask.
+        Args:
```
Comment (wHao-Wu) [inline mmdet3d/datasets/pipelines/loading.py:800]: Also, is it necessary to insert a blank line between the summary and the args?

File: mmdet3d/datasets/pipelines/loading.py [hunk 3]
```diff
@@ -683,3 +684,330 @@ def __repr__(self):
         repr_str += f'{indent_str}with_bbox_depth={self.with_bbox_depth}, '
         repr_str += f'{indent_str}poly2mask={self.poly2mask})'
         return repr_str
+
+
+@PIPELINES.register_module()
+class GenerateEgdeIndices(object):
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:690]: Edge

**Comment by ZCMax** [comment]
According to the discussion, the data label generation has been moved to head module as function format for better implementation

File: mmdet3d/models/utils/edge_indices.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+        edge_indices_edge = np.stack((x, y), axis=1)
+        edge_indices_edge[:, 0] = np.clip(edge_indices_edge[:, 0], x_min,
+                                          x_max)
+        edge_indices_edge[:, 1] = np.clip(edge_indices_edge[:, 1], y_min,
+                                          y_max)
+        edge_indices.append(edge_indices_edge)
+
+        # bottom
+        x = np.arange(x_min, x_max, step)
```
Comment (ZwwWayne) [inline mmdet3d/models/utils/edge_indices.py:47]: considering put it to torch and calculate it with CUDA

File: mmdet3d/models/utils/trunc_objs_handle.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            image, each has shape (num_gt, bbox_code_size).
+        gt_labels_3d_list (list[Tensor]): 3D Ground truth labels of each
+            box, each has shape (num_gt,).
+        centers2d_list (list[Tensor]): Projected 3D centers onto 2D image,
+            each has shape (num_gt, 2).
+        img_metas (list[dict]): Meta information of each image, e.g.,
+            image size, scaling factor, etc.
+
+    Returns:
+        None
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/trunc_objs_handle.py:26]: Can remove

File: mmdet3d/models/utils/edge_indices.py (hunk 4 above)
Comment (ZCMax) [reply mmdet3d/models/utils/edge_indices.py:46]: As measured, the speed of numpy is faster than torch on CUDA due to loop.

Comment (ZCMax) [reply mmdet3d/models/utils/edge_indices.py]: The operation has been moved to torch with CUDA.

File: mmdet3d/models/utils/handle_objs.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        centers2d_list[i] = centers2d[keep_inds]
+        gt_labels_list[i] = gt_labels_list[i][keep_inds]
+        gt_bboxes_list[i] = gt_bboxes_list[i][keep_inds]
+        gt_bboxes_3d_list[i].tensor = gt_bboxes_3d_list[i].tensor[keep_inds]
+        gt_labels_3d_list[i] = gt_labels_3d_list[i][keep_inds]
+
+    return
+
+
+def get_target_centers2d(centers2d, centers, img_shape):
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/handle_objs.py:41]: get_centers2d_target

File: mmdet3d/models/utils/handle_objs.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    inds = (intersects_x >= 0) & (intersects_x <=
+                                  w - 1) & (intersects_y >= 0) & (
+                                      intersects_y <= h - 1)
+    valid_intersects = intersects[inds].reshape(N, 2, 2)
+    dist = torch.norm(
+        valid_intersects - centers2d.unsqueeze(1), dim=2)  # (n, 4)
+    min_idx = torch.argmin(dist, dim=1)  # (n, )
+
+    min_idx = min_idx.unsqueeze(-1).unsqueeze(-1).expand(-1, -1, 2)
+    target_centers2d = valid_intersects.gather(dim=1, index=min_idx).squeeze(1)
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/handle_objs.py:81]: centers2d_target

File: mmdet3d/models/utils/edge_indices.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                     step=1,
+                     pad_mode='default',
+                     dtype=torch.float32,
+                     device='cpu'):
+    """Function to filter the objects label outside the image.
+
+    Args:
+        img_metas (list[dict]): Meta information of each image, e.g.,
+            image size, scaling factor, etc.
+        step (int): Step size used for generateing edge indices.
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/edge_indices.py:15]: int, optional
Please also modify other docstrings

File: mmdet3d/models/utils/edge_indices.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            Default: 1.
+        pad_mode (str): Padding mode during data pipeline.
+            Default: 'default'.
+        dtype (torch.dtype): Dtype of edge indices tensor.
+            Default: torch.float32.
+        device (str): Device of edge indices tensor. Default: 'cpu'.
+
+    Returns:
+        tuple:
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/edge_indices.py:24]: Should be a list?
No need for the variable name here, just revise it as "list[Tensor]: xxxxx"

File: mmdet3d/models/utils/gen_keypoints.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        img_metas (list[dict]): Meta information of each image, e.g.,
+            image size, scaling factor, etc.
+        use_local_coords (bool, optional): Wheher to use local coordinates
+            for keypoints. Default: True.
+
+    Returns:
+        tuple:
+            keypoints2d_list (list[Tensor]): Keypoints for each projected
+                2D bbox.
+            keypoints_depth_mask_list (list[Tensor]): The visible mask of
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/gen_keypoints.py:27]: No need for variable names, please also modify other docstrings

File: mmdet3d/models/utils/handle_objs.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+            (centers2d[:, 0] < img_shape[1]) & \
+            (centers2d[:, 1] > 0) & \
+            (centers2d[:, 1] < img_shape[0])
+        centers2d_list[i] = centers2d[keep_inds]
+        gt_labels_list[i] = gt_labels_list[i][keep_inds]
+        gt_bboxes_list[i] = gt_bboxes_list[i][keep_inds]
+        gt_bboxes_3d_list[i].tensor = gt_bboxes_3d_list[i].tensor[keep_inds]
+        gt_labels_3d_list[i] = gt_labels_3d_list[i][keep_inds]
+
+    return
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/handle_objs.py:38]: Can we remove this return?

File: mmdet3d/models/utils/gen_keypoints.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        centers2d_list (list[Tensor]): Projected 3D centers onto 2D image,
+            shape (num_gt, 2).
+        img_metas (list[dict]): Meta information of each image, e.g.,
+            image size, scaling factor, etc.
+        use_local_coords (bool, optional): Wheher to use local coordinates
+            for keypoints. Default: True.
+
+    Returns:
+        tuple[list[Tensor]]: It contains two elements, the first is the
+        keypoints for each projected 2D bbox in batch data. The second is
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/gen_keypoints.py:25]: Incorrect indent?

File: mmdet3d/models/utils/handle_objs.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        centers2d_list[i] = centers2d[keep_inds]
+        gt_labels_list[i] = gt_labels_list[i][keep_inds]
+        gt_bboxes_list[i] = gt_bboxes_list[i][keep_inds]
+        gt_bboxes_3d_list[i].tensor = gt_bboxes_3d_list[i].tensor[keep_inds]
+        gt_labels_3d_list[i] = gt_labels_3d_list[i][keep_inds]
+
+
+def get_centers2d_target(centers2d, centers, img_shape):
+    """Function to get target centers2d.
+    Args:
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/handle_objs.py:41]: Need a blank line above Args

File: mmdet3d/models/utils/handle_objs.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+        gt_bboxes_list (list[Tensor]): Ground truth bboxes of each image,
+            shape (num_gt, 4).
+        centers2d_list (list[Tensor]): Projected 3D centers onto 2D image,
+            shape (num_gt, 2).
+        img_metas (list[dict]): Meta information of each image, e.g.,
+            image size, scaling factor, etc.
+
+    Returns:
+        tuple[list[Tensor]]: It contains three elements. The first is the
+        target centers2d after handling the truncated objects. The second
```
Comment (Tai-Wang) [inline mmdet3d/models/utils/handle_objs.py:98]: Indent

Comment (Tai-Wang) [reply mmdet3d/models/utils/handle_objs.py:98]: We can check the format in the pre-compiled API documentation.

File: mmdet3d/models/utils/edge_indices.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+import torch
+
+
+def get_edge_indices(img_metas,
+                     step=1,
+                     pad_mode='default',
+                     dtype=np.float32,
+                     device='cpu'):
+    """Function to filter the objects label outside the image.
+    The edge_indices are generated using numpy on cpu, not tensor
```
Comment (ZwwWayne) [inline mmdet3d/models/utils/edge_indices.py:12]: on cpu, not -> on cpu rather than on CUDA due to the latency issue.
When batch size = 8, this function with numpy array is ~8 times faster than that with CUDA tensor (0.09s and 0.72s in 100 runs).
